builtin/logical/aws/rotation.go

```diff
@@ -0,0 +1,143 @@
+			} else {
+				return nil
+			}
+		}
+	}
+}
+
+// rotateCredential pops an element from the priority queue, and if it is expired, rotate and re-push.
+// If a cred was rotated, it returns true, otherwise false.
+func (b *backend) rotateCredential(ctx context.Context, storage logical.Storage) (rotated bool, err error) {
+	// If queue is empty or first item does not need a rotation (priority is next rotation timestamp) there is nothing to do
+	item, err := b.credRotationQueue.Pop()
+	if err != nil {
+		// the queue is just empty, which is fine.
+		if err == queue.ErrEmpty {
+			return false, nil
+		}
+		return false, fmt.Errorf("failed to pop from queue for role %q: %w", item.Key, err)
+	}
+	if item.Priority > time.Now().Unix() {
+		// no rotation required
+		// push the item back into priority queue
+		err = b.credRotationQueue.Push(item)
+		if err != nil {
+			return false, fmt.Errorf("failed to add item into the rotation queue for role %q: %w", item.Key, err)
+		}
+		return false, nil
+	}
+
+	cfg := item.Value.(staticRoleEntry)
+
+	err = b.createCredential(ctx, storage, cfg)
+	if err != nil {
+		return false, err
+	}
+
+	// set new priority and re-queue
+	item.Priority = time.Now().Add(cfg.RotationPeriod).Unix()
+	err = b.credRotationQueue.Push(item)
+	if err != nil {
+		return false, fmt.Errorf("failed to add item into the rotation queue for role %q: %w", cfg.Name, err)
+	}
+
+	return true, nil
+}
+
+func (b *backend) createCredential(ctx context.Context, storage logical.Storage, cfg staticRoleEntry) error {
+	iamClient, err := b.clientIAM(ctx, storage)
+	if err != nil {
+		return fmt.Errorf("unable to get the AWS IAM client: %w", err)
+	}
+
+	// IAM users can have a most 2 sets of keys at a time.
+	// (https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-quotas.html)
+	// Ideally we would get this value through an api check, but I'm not sure one exists.
+	const maxAllowedKeys = 2
+
+	accessKeys, err := iamClient.ListAccessKeys(&iam.ListAccessKeysInput{
+		UserName: aws.String(cfg.Username),
+	})
+	if err != nil {
+		return fmt.Errorf("unable to list existing access keys for IAM user '%s': %w", cfg.Username, err)
+	}
+
+	// If we have the maximum number of keys, we have to delete one to make another (so we can get the credentials).
+	// We'll delete the oldest one.
+	//
+	// Since this check relies on a pre-coded maximum, it's a bit fragile. If the number goes up, we risk deleting
+	// a key when we didn't need to. If this number goes down, we'll start throwing errors because we think we're
+	// allowed to create a key and aren't. In either case, adjusting the constant should be sufficient to fix things.
+	if len(accessKeys.AccessKeyMetadata) >= maxAllowedKeys {
+		oldestKey := accessKeys.AccessKeyMetadata[0]
+
+		for i := 1; i < len(accessKeys.AccessKeyMetadata); i++ {
+			if accessKeys.AccessKeyMetadata[i].CreateDate.Before(*oldestKey.CreateDate) {
+				oldestKey = accessKeys.AccessKeyMetadata[1]
```

Comment:
Shouldn't the index here be `i` instead of 1?

Reply:
yep! missed a line when updating

Reply:
I wrote up a test for `createCredential`, but it's a little difficult to test what we actually want to (i.e., whether we're trying to delete the correct key), because AFAICT, `awsutil.MockIAM` doesn't provide an obvious way to inspect how we called it.